enzyme/Enzyme/MLIR/Analysis/ActivityAnnotations.cpp

```diff
@@ -0,0 +1,1319 @@
+#include "ActivityAnnotations.h"
+#include "DataFlowAliasAnalysis.h"
+#include "DataFlowLattice.h"
+#include "Dialect/Dialect.h"
+#include "Dialect/Ops.h"
+
+#include "mlir/Analysis/DataFlow/ConstantPropagationAnalysis.h"
+#include "mlir/Analysis/DataFlow/DeadCodeAnalysis.h"
+#include "mlir/Analysis/DataFlow/DenseAnalysis.h"
+#include "mlir/Analysis/DataFlow/SparseAnalysis.h"
+#include "mlir/Analysis/DataFlowFramework.h"
+#include "mlir/Interfaces/FunctionInterfaces.h"
+#include "llvm/Support/raw_ostream.h"
+
+// TODO: Remove dependency on dialects in favour of differential dependency
+// interface
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+
+using namespace mlir;
+
+static bool isPossiblyActive(Type type) {
+  return isa<FloatType, ComplexType>(type);
+}
+
+template <typename ValueT>
+void printSetLattice(const enzyme::SparseSetLattice<ValueT> &setLattice,
+                     raw_ostream &os) {
+  if (setLattice.isUnknown()) {
+    os << "Unknown Origin";
+  } else if (setLattice.isUndefined()) {
+    os << "Undefined Origin";
+  } else {
+    os << "size: " << setLattice.getElements().size() << ":\n";
+    for (auto element : setLattice.getElements()) {
+      os << "  " << element << "\n";
+    }
+  }
+}
+
+void enzyme::ForwardOriginsLattice::print(raw_ostream &os) const {
+  printSetLattice(*this, os);
+}
+
+void enzyme::BackwardOriginsLattice::print(raw_ostream &os) const {
+  printSetLattice(*this, os);
+}
+
+ChangeResult
+enzyme::ForwardOriginsLattice::join(const AbstractSparseLattice &other) {
+  const auto *otherValueOrigins =
+      static_cast<const ForwardOriginsLattice *>(&other);
+  return elements.join(otherValueOrigins->elements);
+}
+
+void enzyme::ForwardActivityAnnotationAnalysis::setToEntryState(
+    ForwardOriginsLattice *lattice) {
+  auto arg = dyn_cast<BlockArgument>(lattice->getAnchor());
+  if (!arg) {
+    assert(lattice->isUndefined());
+    return;
+  }
+  if (!isPossiblyActive(arg.getType())) {
+    return;
+  }
+
+  auto funcOp = cast<FunctionOpInterface>(arg.getOwner()->getParentOp());
+  auto origin = ArgumentOriginAttr::get(FlatSymbolRefAttr::get(funcOp),
+                                        arg.getArgNumber());
+  return propagateIfChanged(
+      lattice, lattice->join(ForwardOriginsLattice::single(lattice->getAnchor(),
+                                                           origin)));
+}
+
+/// True iff all results differentially depend on all operands
+// TODO: differential dependency/activity interface
```

Comment:
So we do now have such an interface: https://github.com/EnzymeAD/Enzyme/blob/716d674d012834e6a1da0ad5febf9194e20aa868/enzyme/Enzyme/MLIR/Analysis/ActivityAnalysis.cpp#L1001